File: Problems/1-100/33.SearchinRotatedSortedArray/SearchInRotatedSortedArray.java
public class SearchInRotatedSortedArray {
    /**
     * binary search
     * O(logn)
     * @param nums
     * @param target
     * @return
     */
    public int search1(int[] nums, int target) {
        if(target == nums[0])
            return 0;
        if(target == nums[nums.length-1])
            return nums.length-1;
        int l = 0;
        int r = nums.length-1;
        while(l<=r){
            int mid = (l+r) /2;
            if(nums[mid] == target ){
                return mid;
            }else if (target > nums[mid]){
                if(nums[mid]>= nums[l]){
                    l = mid+1;
                }else{
                    if(target >= nums[l]){
                        r = mid-1;
                    }else{
                        l = mid +1;
                    }
                }
            }else{
                if(nums[mid] >= nums[l]){
                    if(target < nums[l]){
                        l = mid+1;
                    }else{
                        r = mid-1;
                    }
                }else{
                    r= mid-1;
                }

            }
        }
        return -1;
    }
    /**
     * 判断从前往后还是从后往前
     * O(n/2) 即 O(N)
     * @param nums
     * @param target
     * @return
     */
    public int search(int[] nums, int target) {
        if(target == nums[0])
            return 0;
        if(target == nums[nums.length-1])
            return nums.length-1;
        if(target > nums[0]){
            for(int i = 1; i< nums.length && nums[i] >= nums[i-1]; i++){
                if(nums[i] == target)
                    return i;
            }
        }else{
            for(int i = nums.length-2; i >=0 && nums[i] <= nums[i+1]; i--){
                if(nums[i] == target)
                    return i;
            }
        }
        return -1;
    }
}

File: past/search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
       
        l = 0
        r = len(nums) -1 
        while r >= l:
            if r == l :
                return -1 if nums[l] != target else 0
            elif nums[l] < nums[r]:
                return self.regularsearch(nums, target, l, r)
            elif r-l == 1:
                if nums[l] == target:
                    return l
                elif nums[r] == target:
                    return r
                else:
                    return -1
            else:
                mid = (r+l) //2
                if (nums[mid] == target):
                    return mid
                else:
                    leftpart_r = mid-1
                    rightpart_l = mid + 1
                    #border on the left
                    if nums[l] > nums[leftpart_r]:
                        if target >= nums[l] or target <= nums[leftpart_r]:
                            r = leftpart_r
                            
                        else:
                            l = rightpart_l
                            return self.regularsearch(nums, target,l, r)
                    #border on the right
                    elif nums[rightpart_l] > nums[r]:
                        if target <= nums[r] or target >= nums[rightpart_l]:
                            l = rightpart_l

                        else:
                            r = leftpart_r
                            return self.regularsearch(nums,target, l, r)
                    else:
                        if target > nums[r]:
                            r = leftpart_r
                            return self.regularsearch(nums,target, l ,r)
                        else:
                            l = rightpart_l
                            return self.regularsearch(nums,target,l ,r)

                            
        return -1
    def regularsearch(self, nums, target, l, r):
        while r >= l:
            mid = (r+l) //2
            if nums[mid] == target:
                return mid
            if nums[mid] > target:
                r = mid-1
            else:
                l = mid + 1
        return -1
                        
    
        